src/orion/benchmark/base.py

```diff
@@ -0,0 +1,116 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+"""
+:mod:`orion.benchmark.base` -- Base definition of Task and Assessment
+======================================================================
+
+.. module:: base
+   :platform: Unix
+   :synopsis: Base definition of Task and Assessment.
+
+"""
+
+from abc import ABC, abstractmethod
+
+
+class BaseAssess(ABC):
+    """Base class describing what an assessment can do.
+
+    Parameters
+    ----------
+    task_num : int
+       Number of experiment the assessment ask to run the corresponding task
+    kwargs : dict
+       Configurable parameters of the assessment, a particular assessment
+       implementation can have its own parameters.
+    """
+
+    def __init__(self, task_num, **kwargs):
+        self.task_number = task_num
+        self._param_names = list(kwargs.keys())
+
+    @property
+    def task_num(self):
+        """Return the task number to run for this assessment"""
+        return self.task_number
+
+    @abstractmethod
+    def plot_figures(self, task, experiments):
+        """
+        Generate a `plotly.graph_objects.Figure`
+
+        task: str
+            Name of the task
+        experiments: list
+            A list of (task_index, experiment), where task_index is the index of task to run for
+            this assessment, and experiment is an instance of `orion.core.worker.experiment`.
+        """
+        pass
+
+    @property
+    def configuration(self):
+        """Return the configuration of the assessment."""
+        dict_form = dict()
+        for attrname in self._param_names:
+            if attrname.startswith("_"):  # Do not log _space or others in conf
+                continue
+            attr = getattr(self, attrname)
+            dict_form[attrname] = attr
+        dict_form["task_num"] = self.task_num
+
+        mod = self.__class__.__module__
+        fullname = mod + "." + self.__class__.__qualname__
+        fullname = fullname.replace(".", "-")
+        return {fullname: dict_form}
+
+
+class BaseTask(ABC):
+    """Base class describing what a task can do.
+    A task will define the objective function and search space of it.
+
+    Parameters
+    ----------
+    max_trials : int
+       Max number of trials the experiment will run against this task.
+    kwargs : dict
+       Configurable parameters of the task, a particular task
+       implementation can have its own parameters.
+    """
+
+    def __init__(self, max_trials, **kwargs):
+        self.trials_num = max_trials
+        self._param_names = list(kwargs.keys())
+
+    @abstractmethod
+    def get_blackbox_function(self):
```

Comment:
Do you see an advantage of returning a function instead of defining it as ```__call__``` so that we can directly work on the task object?

Reply:
`__call__` looks a more straightforward way, but I am thinking how to guarantee the sub-task implementation will always get this `__call__`. What's your suggestion?

Reply:
another possible way is that we add `__call__` into the base task and replace this `get_black_box_function` with a `call`, and in the `__call__`, we call the sub-class `call()`.

Reply:
Can we make the `__call__` an abstract method? Otherwise I think we should make it raise `NotImplemented` error so that subclasses that fails to implement it fails in an obvious way.

Reply:
I guess we can. But abstract `__call__` seems a little bit weird for me. That's why I suggest some way else in the last comment.

Reply:
I think it's reasonable to force implementation of ` __call__` in children. Do you find it weird because it's an operator function?

Reply:
yeah, it is the python magic methods, for me, it is more like some reserved methods. 

Some possible reference like, TensorFlow is asking `call()` and Pytorch is asking `forward` in the sub class for similar purpose.

https://www.tensorflow.org/tutorials/customization/custom_layers#implementing_custom_layers
https://pytorch.org/tutorials/beginner/examples_nn/two_layer_net_module.html

Reply:
Sounds good, let's go with `call()` as you suggested.